### PR TITLE
Enhancing DisjointSetUnion data structure

### PR DESCRIPTION
**Pull request by Manan-09**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

Done enhancing of DisjoinSetUnion data structure code, added comments , unit testcase and updated method name. Also updated package name to disjointsets -> disjointsetunion .
Screenshot 2023-09-10 at 5 56 30 PM

### CODE REVIEW

**Comment by Manan-09** [comment]
Please help with this lint error , what I need to do ? 

```
@@ -22,4 +22,4 @@
         this.data = data;
         parent = this; // Initially, a node is its own parent.
     }
-}+}
--- ./src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java	(original)
+++ ./src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java	(reformatted)
@@ -50,4 +50,4 @@
             nx.rank++;
         }
     }
-}+}
```

**Comment by BamaCharanChhandogi** [comment]
You have to give one space before the curly bracket.

**Comment by BamaCharanChhandogi** [comment]
> Please help with this lint error , what I need to do ?
> 
> ```
> @@ -22,4 +22,4 @@
>          this.data = data;
>          parent = this; // Initially, a node is its own parent.
>      }
> -}+}
> --- ./src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java	(original)
> +++ ./src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java	(reformatted)
> @@ -50,4 +50,4 @@
>              nx.rank++;
>          }
>      }
> -}+}
> ```

You have to give one space before the curly bracket.

**Comment by Manan-09** [comment]
Thanks @BamaCharanChhandogi

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        Node<T> nx = findSet(x);
+        Node<T> ny = findSet(y);
+
+        if (nx == ny) {
+            return; // Both elements already belong to the same set.
+        }
+        // Merging happens based on rank of node, this is done to avoid long chaining of nodes and reduce time
+        // to find root of the component. Idea is to attach small components in big, instead of other way around.
+        if (nx.rank > ny.rank) {
+            ny.parent = nx;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java:44]: This line is not covered by any test.

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        } else if (ny.rank > nx.rank) {
+            nx.parent = ny;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java:46]: This line is not covered by any test.

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * Unions two sets by merging their representative elements. The merge is performed based on the rank of each set
+     * to ensure efficient merging and path compression to optimize future findSet operations.
+     */
+    public void unionSets(Node<T> x, Node<T> y) {
+        Node<T> nx = findSet(x);
+        Node<T> ny = findSet(y);
+
+        if (nx == ny) {
+            return; // Both elements already belong to the same set.
```
Comment (vil02) [inline src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java:39]: This line is not covered by any test.

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/Node.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+     * Initially, a node is its own parent (represents a singleton set).
+     */
+    public Node<T> parent;
+
+    /**
+     * The data element associated with the node.
+     */
+    public T data;
+
+    public Node(T data) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/datastructures/disjointsetunion/Node.java:21]:
```suggestion
    public Node(final T data) {
```

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * It is commonly employed in graph algorithms and problems.
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Disjoint-set_data_structure">Disjoint Set Union</a>
+ */
+public class DisjointSetUnion<T> {
+
+    /**
+     * Creates a new node of DSU with parent initialised as same node
+     */
+    public Node<T> makeSet(T x) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java:15]:
```suggestion
    public Node<T> makeSet(final T x) {
```

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            node = node.parent;
+        }
+        return node;
+    }
+
+    /**
+     * Unions two sets by merging their representative elements. The merge is performed based on the rank of each set
+     * to ensure efficient merging and path compression to optimize future findSet operations.
+     */
+    public void unionSets(Node<T> x, Node<T> y) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java:34]:
```suggestion
    public void unionSets(final Node<T> x, final Node<T> y) {
```

**Review by vil02 — CHANGES_REQUESTED** [review]
Some quick remarks form my side.

When I will find some nice test covering the missing lines, I will put it here know.

File: src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java (hunk 6 above)
Comment (Manan-09) [reply src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java]: Updated testcases accordingly to check all branches
Screenshot 2023-09-10 at 9 38 14 PM

**Comment by Manan-09** [comment]
> Some quick remarks form my side.
> 
> When I will find some nice test covering the missing lines, I will put it here know.
Screenshot 2023-09-10 at 9 38 14 PM
@vil02  Updated testcases, covering all if branches

**Review by vil02 — APPROVED** [review]

**Pull request merged**